### PR TITLE
Don't update collection initializer statement that references collection being created

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/18260
Fixes https://github.com/dotnet/roslyn/issues/18271

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by sharwell** [comment]
:bulb: It would be good to add the specific test cases seen in the bugs this claims to fix.

:bulb: Can you also add the following test case?

```csharp
using System.Collections.Generic;

class C
{
    private List<int> myField;
    void M()
    {
        myField = new List<int>();
        myField.Add(this.myField.Count);
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Have added the tests.  Also refactored things to allow more code to be shared between use-object-initializer and use-collection-initializer so that we don't keep having to make duplicate changes to both.

File: src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs [hunk 1]
```diff
@@ -197,95 +186,6 @@ private bool ImplicitMemberAccessWouldBeAffected(SyntaxNode node)
 
             return false;
         }
-
-        private bool ExpressionContainsValuePattern(TExpressionSyntax expression)
```
Comment (sharwell) [inline src/Features/Core/Portable/UseObjectInitializer/ObjectCreationExpressionAnalyzer.cs]: :memo: I'm assuming this is a trivial move to the base class, so I didn't review for preservation of details in this specific refactoring. This applies to `UseCollectionInitializer` case as well.

**Pull request merged**